File: solutions/zig/nucleotide-count/1/nucleotide_count.zig
pub const NucleotideError = error{Invalid};

pub const Counts = struct { a: u32, c: u32, g: u32, t: u32 };

pub fn countNucleotides(s: []const u8) NucleotideError!Counts {
    var c: Counts = .{ .a = 0, .c = 0, .g = 0, .t = 0 };
    for (s) |n| switch (n) {
        'A' => c.a += 1,
        'C' => c.c += 1,
        'G' => c.g += 1,
        'T' => c.t += 1,
        else => return NucleotideError.Invalid,
    };
    return c;
}

File: solutions/go/nucleotide-count/1/nucleotide_count.go
package dna

import "errors"

// Histogram is a mapping from nucleotide to its count in given DNA.
type Histogram map[rune]int

// DNA is a list of nucleotides.
type DNA string

func (d DNA) Counts() (Histogram, error) {
	h := Histogram{'A': 0, 'C': 0, 'G': 0, 'T': 0}
	for _, n := range d {
		if _, ok := h[n]; !ok {
			return nil, errors.New("invalid nucleotide")
		}
		h[n]++
	}
	return h, nil
}
